### PR TITLE
TensorFlow vs PyTorch: Which one to pick?

### LINKED ISSUE #783: TensorFlow vs PyTorch: Which one to pick?
### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.

TensorFlow and PyTorch are both deep learning frameworks used actively for research and development of deep learning application. In this article, we'll be exploring which one is the best to use and why? 

### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?

By the end of this article, the reader should understand:

1. What is TensorFlow?
2. What is PyTorch?
3. What are the differences between these two frameworks?
4. Which one to pick?

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by Bayler**
Fixes #783

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

File: articles/tensorflow-vs-pytorch/index.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+### Prerequisites
+
+Deep Learning (DL) frameworks all center around deep learning and neural networks. If you haven't studied neural networks and how they function, please feel free to read this [article](https://www.section.io/engineering-education/differences-between-artificial-intelligence-machine-learning-and-deep-learning/) before diving into reading this article.
```
Comment (lalith1403) [inline articles/tensorflow-vs-pytorch/index.md:3]: These frameworks don't center around neural networks. They are gradient computing engines. Modify your statement.

File: articles/tensorflow-vs-pytorch/index.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+### What is TensorFlow (TF)?
+
+TensorFlow is a software application popular for implementing deep learning algorithms. The Google Brain team developed it, and it was released as an open-source platform in 2015. It supports both the R and Python programming languages. It was initially started as a research-based architecture for Google's in-house applications but was eventually open-sourced to be used commercially.
+
+It is known as "TensorFlow" because it takes in inputs as tensors or multi-dimensional arrays (MDAs). These tensors (MDAs) then flow through a system of operations and comes out the other end as outputs.
+
+The key features of TensorFlow include:
+
+1. It is actively maintained by the developers at Google. Currently, TF2.3 is the latest released version.
```
Comment (lalith1403) [inline articles/tensorflow-vs-pytorch/index.md:20]: Add link to tf2.3

File: articles/tensorflow-vs-pytorch/index.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+### What is PyTorch?
+
+PyTorch is a relatively new deep learning framework as it came out a year later after TensorFlow was released. The name PyTorch was inspired by a popular "[torch](http://torch.ch/)" deep learning framework, which was written in the [Lua](https://www.lua.org/) programming language. However, learning the Lua programming language was a big barrier, especially when learning deep learning. It also didn't offer the modularity necessary to interface with other deep learning libraries currently present. 
+
+This problem was solved by a team of AI researchers inspired by the torch's programming style. They decided to develop an improvised form of torch with the Python (py) programming language, which later became known as PyTorch. 
+
+Some notable features of PyTorch include:
+
+1. It is extremely fast and feels Native.
```
Comment (lalith1403) [inline articles/tensorflow-vs-pytorch/index.md:35]: Remove unnecessary capitalization.

File: articles/tensorflow-vs-pytorch/index.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+2. It has native support for Python and all of its libraries. This feature makes it very easy to learn. 
+3. It's easy to use API that ensures easy usability when making use of the API.
+4. It is actively maintained and used by the developers at Facebook.
+5. It supports Nvidia's [CUDA](https://en.wikipedia.org/wiki/CUDA), ensuring that the code runs on a GPU, decreasing the time needed to run the code.
+6. Dynamic Computation Graphs is a key feature in PyTorch. It ensures that the graphs build up dynamically as you code.
```
Comment (lalith1403) [inline articles/tensorflow-vs-pytorch/index.md:40]: Explain what a dynamic computation graph is. Add a section on static graphs vs. dynamic graphs and how it affects the performance. This section forms the heart of this article.

File: articles/tensorflow-vs-pytorch/index.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+By default, when [TensorFlow](https://www.tensorflow.org/) was developed, it let developers create static computation graphs at compile time. It didn't allow us to define and manipulate graphs at runtime. If we wanted to make any changes, we would have to rebuild the neural network from the start. 
+
+But, with the release of TensorFlow 2.0, they introduced the "eager mode." They aimed to allow graphs to be defined and manipulated dynamically in computation while executing as was with PyTorch, which already had this feature, which worked so well. Rather than waiting until everything is coded up and connected before you can see if anything is done right, you could now "eagerly" execute your operations as you're working. 
+
+Today, both frameworks allow both the static and dynamic computation of graphs. No much difference between the two as far as computation graphs are concerned.
```
Comment (lalith1403) [inline articles/tensorflow-vs-pytorch/index.md:54]: Rewrite the last sentence. As far as computational graphs are concerned, not much difference exists between the two.

File: articles/tensorflow-vs-pytorch/index.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+TensorFlow comes on-board with the [TensorBoard](https://www.tensorflow.org/tensorboard) visualization toolkit that helps researchers understand graphs and how their code runs. It is also an easy-to-use toolkit that helps track and visualize metrics such as accuracy and losses in your model. 
+
+PyTorch does not have a native in-build visualizing tool like TensorFlow does. A researcher has to import external visualization libraries such as [matplotlib](https://matplotlib.org/) and [seaborn](https://seaborn.pydata.org/) to be able to achieve the same objective.
+
+They both achieve the same visualization objective, but TensorBoard offers researchers a bigger advantage. It has more features and a good number of visualization tools that transform graphs into beautiful-looking visualizations.
+
+### Community
+
+As I mentioned earlier, TensorFlow has been around for a longer time compared to PyTorch. Due to this factor, and partly because Google's technology giants develop it, TensorFlow has a much wider community around its ecosystem. This community helps make the learning of the framework easier.
```
Comment (lalith1403) [inline articles/tensorflow-vs-pytorch/index.md:68]: This statement is actually wrong. The research community has embraced PyTorch really well. Therefore, I suggest you do some research and update this statement. PyTorch has a solid community presence.

**Review by lalith1403 — CHANGES_REQUESTED** [review]

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by lalith1403 — APPROVED** [review]
